src/shared/parameter_v2/generation_scripts/cpp_config.py

```diff
@@ -0,0 +1,413 @@
+    MutableParameterList mutable_internal_param_list;
+    ParameterList immutable_internal_param_list;
+    {private_entries}
+}};
+"""
+
+
+class CppConfig(object):
+    def __init__(self, config_name: str, is_top_level_config: bool = False):
+        self.config_name = config_name
+        self.config_variable_name = to_snake_case(config_name) + "_config"
+        self.is_top_level_config = is_top_level_config
+        self.parameters: List[CppParameter] = list()
+        self.configs: List[CppConfig] = list()
+
+    def add_parameter(self, parameter: CppParameter):
+        self.parameters.append(parameter)
+
+    def get_parameters(self) -> List[CppParameter]:
+        return self.parameters.copy()
+
+    def include_config(self, config: CppConfig):
+        self.configs.append(config)
+
+    @staticmethod
+    def join_with_tabs(
+        str: str, iterable: Iterable, num_tabs: int, half_tab: bool = False
+    ) -> str:
+        tab_size = HALF_TAB_SIZE if half_tab else TAB_SIZE
+        return (str + num_tabs * tab_size * " ").join(iterable)
+
+    @property
+    def dependency_graph(self):
+        return self.__dependency_graph
+
+    @dependency_graph.setter
+    def dependency_graph(self, dependency_graph: nx.DiGraph):
+        self.__dependency_graph = dependency_graph
+
+        self.dependency_graph_topological_order_configs = [
+            self.dependency_graph.nodes[node]["config"]
+            for node in list(reversed(list(nx.topological_sort(self.dependency_graph))))
+        ]
+
+        self.included_config_command_line_arg_entries = []
+        self.included_config_load_command_line_args_into_config_contents = []
+        for config in self.configs:
+            self.dfs_helper(config, "", "")
+
+    def dfs_helper(self, config: CppConfig, arg_prefix: str, load_dependency: str):
+        arg_prefix = (
+            to_snake_case(config.config_name)
+            if not arg_prefix
+            else arg_prefix + "." + to_snake_case(config.config_name)
+        )
+        load_dependency = load_dependency + "getMutable{config_name}()->".format(
+            config_name=config.config_name
+        )
+
+        mutable_param_gen = (
+            param for param in config.parameters if not param.is_constant
+        )
+        for param in mutable_param_gen:
+            self.included_config_command_line_arg_entries.append(
+                param.command_line_option_entry_with_prefix(arg_prefix + ".")
+            )
+            self.included_config_load_command_line_args_into_config_contents.append(
+                param.load_command_line_arg_into_config_with_dependencies(
+                    load_dependency, arg_prefix + "."
+                )
+            )
+
+        # top level config has access to all configs, so no need to recursively add options
+        # if not self.is_top_level_config:
+        for included_config in config.configs:
+            self.dfs_helper(included_config, arg_prefix, load_dependency)
```

Comment:
This is a little bugged, I forgot to add in some `constant` checks for some of these lines, will fix them. `foo_bool` and `foo_int` is correct not being there though since they are marked `constant`, and with how they are initialized, they can't be loaded in from command line.

Reply:
Fixed - now only non constant parameters should show up in the `loadFromCommandLineArguments` function. If it is needed to be able to load constant parameters via command line args, I could also add that later.